feat: support github deploy statuses for previews

mikestopcontinues opened the pull request:
Fixes #4359

**Summary**

This PR adds github's deployment statuses in addition to direct commit statuses. Using `preview_context`, you can now select from matching deployments.

**Test plan**

The test is a little involved, but it works.

Here's the backend config:

```yaml
backend:
  name: github
  repo: mikestopcontinues/netlify-test
  branch: master
  open_authoring: true
  use_graphql: false # also works as true
  preview_context: 'Deployment has completed'
```

Here's the matching [repo](https://github.com/mikestopcontinues/netlify-test) and the [test PR](https://github.com/mikestopcontinues/netlify-test/pull/1).

Here's the PR deployment status Vercel adds (rather than a commit status):

<img width="1334" alt="Screen Shot 2020-10-06 at 6 57 28 PM" src="https://user-images.githubusercontent.com/150434/95272429-df59e000-0805-11eb-9dc1-3ef0debdf7a7.png">

And the proof... Here's the correct preview url. The results are identical when using graphql.

<img width="1239" alt="Screen Shot 2020-10-06 at 6 53 56 PM" src="https://user-images.githubusercontent.com/150434/95272472-02848f80-0806-11eb-9ac8-64b9f2b4ad0d.png">

**A picture of a cute animal (not mandatory but encouraged)**

![YAWWWWWN](https://media3.giphy.com/media/UvIApNhjzIEhi/giphy.gif?cid=ecf05e47pym8zccxx2kf4cbycqfmp5qhp4i37trzx56ltjot&rid=giphy.gif)

erezrokah commented:
Thanks @mikestopcontinues, always nice seeing people close their own issues.
You can test locally by updating the configuration here:
https://github.com/netlify/netlify-cms/blob/576e4f0f1a158d6b587587c52fb288d8f6eea89f/dev-test/config.yml

and running `yarn && yarn start`

erezrokah commented on packages/netlify-cms-backend-github/src/API.ts:
```diff
@@ -864,13 +872,33 @@ export default class API {
     const resp: { statuses: GitHubCommitStatus[] } = await this.request(
       `${this.originRepoURL}/commits/${sha}/status`,
     );
+    const deploys: GitHubDeployments = await this.request(
```
Can we do this only if `statuses.length <= 0`?

mikestopcontinues replied:
Hey, I implemented it this way to work with `preview_context`, so the end user just looks up the related message to filter on. I think using deploy statuses as a fallback (or commit statuses) would require another config option, since it's entirely possible the user could have different bots leaving messages in both places for different reasons. I don't like the extra requests any more than you do though. I'm open to changing it, or maybe even to just removing the feature from the non-gql API, since I'm assuming it will eventually be retired. What do you think?

erezrokah replied:
We can't retire the REST API since it is also used by the `git-gateway` backend.
I does makes sense to check both.

erezrokah commented on packages/netlify-cms-backend-github/src/queries.ts:
```diff
@@ -54,6 +54,16 @@ export const statues = gql`
               target_url: targetUrl
             }
           }
+          deployments(last: 1) {
```
Nice!

erezrokah commented on packages/netlify-cms-backend-github/src/GraphQLAPI.ts:
```diff
@@ -329,8 +329,11 @@ export default class GraphQLAPI extends API {
-      const { status } = data.repository.object;
+      const { status, deployments } = data.repository.object;
       const { contexts } = status || { contexts: [] };
+      if (deployments.nodes[0] && deployments.nodes[0].latestStatus) {
```
Similar comment as before, but less critical here since we don't need to send additional requests.
Does it make sense to extract the status both from the commit and the deployment?

erezrokah commented:
Looks like CI is failing due to formatting issues. You can run https://github.com/netlify/netlify-cms/blob/master/CONTRIBUTING.md#test locally to debug.

mikestopcontinues commented:
> Looks like CI is failing due to formatting issues. You can run https://github.com/netlify/netlify-cms/blob/master/CONTRIBUTING.md#test locally to debug.

I'm having trouble getting the linting to run properly. Prettier is throwing an error with no context (failing during `widget-markdown`). When run separately, the tests pass and all the linting works on `backend-github`. In any case, I'll try to get some proof-of-tests to you ASAP.

erezrokah commented on packages/netlify-cms-backend-github/src/API.ts:
```diff
@@ -91,6 +93,12 @@ type GitHubCommitStatus = Octokit.ReposListStatusesForRefResponseItem & {
 
+type ParsedStatus = {
+  context: string;
+  state: string;
```
`state` should be of type `PreviewState` (type check is currently failing)

erezrokah commented on packages/netlify-cms-backend-github/src/API.ts:
```diff
@@ -864,13 +872,46 @@ export default class API {
+        );
+
+        return stats.reduce(
+          (acc, stat, i) =>
```
Is the purpose of this to return the last successful status or the first one if there is no successful status?

I wonder if something like this would be more clear:
```js
  stats.reverse().forEach(s => {
    if (s.state === GitHubCommitStatusState.Success) {
      return ...
    }
  });

  if (stats.length > 0) {
    return ...
  }
  
  return ...
```

mikestopcontinues replied:
In the process of simplifying the type issue, I split the map/reduce, making it much easier to read.

erezrokah reviewed (changes requested):
Added a few comments @mikestopcontinues.

For linting - maybe try `yarn install -force` to make sure you have the correct versions locally?

mikestopcontinues commented:
I added the test info to the original post, and I fixed the type issue. When it comes to linting, even after `yarn install --force`, prettier was still failing, so I simply ran it manually on the github-backend package. It made some changes, but the CI is still failing.

Here's what happens when I run `run-p -c --aggregate-output "lint:* --quiet"` locally. It's got slightly more info than on the build machine. 

```
> @ lint:css /Users/msc/Code/netlify-cms
> stylelint --ignore-path .gitignore "{packages/**/*.{css,js,jsx,ts,tsx},website/**/*.css}"

Build Package: netlify-cms
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! @ lint:format: `prettier "{{packages,scripts,website}/**/,}*.{js,jsx,ts,tsx,css}" --list-different`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the @ lint:format script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/msc/.npm/_logs/2020-10-07T00_11_56_406Z-debug.log

> @ lint:format /Users/msc/Code/netlify-cms
> prettier "{{packages,scripts,website}/**/,}*.{js,jsx,ts,tsx,css}" --list-different

packages/netlify-cms-backend-github/src/API.ts
packages/netlify-cms-backend-github/src/AuthenticationPage.js
packages/netlify-cms-backend-github/src/GraphQLAPI.ts
packages/netlify-cms-backend-github/src/implementation.tsx

> @ lint:js /Users/msc/Code/netlify-cms
> eslint --color --ignore-path .gitignore "{{packages,scripts,website}/**/,}*.{js,jsx,ts,tsx}"

Build Package: netlify-cms
ERROR: "lint:format --quiet" exited with 1.
```

erezrokah commented on packages/netlify-cms-backend-github/src/API.ts:
```diff
@@ -864,13 +866,37 @@ export default class API {
+          `${this.originRepoURL}/deployments/${dep.id}/statuses`,
+        );
+
+        return stats
+          .map(stat => ({
+            context: stat.description,
+            // eslint-disable-next-line @typescript-eslint/camelcase
+            target_url: stat.target_url,
+            state:
+              stat.state === GitHubCommitStatusState.Success
+                ? PreviewState.Success
+                : PreviewState.Other,
+          }))
+          .reduce((acc, stat) => (stat.state === PreviewState.Success ? stat : acc));
```
```suggestion
        const statuses = stats.map(stat => ({
          context: stat.description,
          // eslint-disable-next-line @typescript-eslint/camelcase
          target_url: stat.target_url,
          state:
            stat.state === GitHubCommitStatusState.Success
              ? PreviewState.Success
              : PreviewState.Other,
        }));

        return statuses.reverse().find(s => s.state === PreviewState.Success) || statuses[0];
```

Would this be easier to read?

erezrokah reviewed (changes requested):
@mikestopcontinues I'm not sure what's happening with the formatting, as when I'm running it locally it matches our CI.
I added another comment on the usage of `reduce` - I think avoiding it can make the code a bit clearer, WDYT?

Also can you share how to set these statuses via a GitHub API call so I can test it?

mikestopcontinues commented:
I'm fully ok with the style change. I've been very into FP these days, so what I'd written was just a sloppy in-between anyway. I'll get back to you with the updated test and an API request sometime this week.

erezrokah commented:
I would love to move this forward. I'm missing an example how to set these statuses via a GitHub API call

erezrokah commented:
Closing this for now. Please comment if still relevant and I'll re-open

elesime commented:
@erezrokah Hi, would it be possible to reopen this?

This is needed to support deploy previews on Vercel.

I don't see a preview link for my posts in Netlify CMS, however, if I inspect the network call to GitHub's statuses API I see the following being returned:

```
{
  "state": "success",
  "statuses": [
    {
      "url": "https://api.github.com/repos/<username>/<repo>/statuses/xxxxxxxxxxxxxxxxxx",
      "avatar_url": "https://avatars.githubusercontent.com/in/xxxxxx?v=4",
      "id": xxxxxxxxxx,
      "node_id": "xxxxxxxxxxx",
      "state": "success",
      "description": "Deployment has completed",
      "target_url": "https://vercel.com/<username>/<repoZ/xxxxxxxxxxxxxxxxxx",
      "context": "Vercel",
      "created_at": "2022-07-17T10:03:42Z",
      "updated_at": "2022-07-17T10:03:42Z"
    }
  ],
  "sha": "xxxxxxxxxxxxxxxxxxxxxxxxxxxx",
  "total_count": 1,
  ...
```

I belive this PR will support my usecase (deployments on Vercel). Let me know if I can help in any way.

Thank you!

ejmuentes commented:
> I would love to move this forward. I'm missing an example how to set these statuses via a GitHub API call

Example of how to set the statuses and create deployments is here:
https://github.com/netlify/netlify-cms/pull/5733#issuecomment-902811156

I am using the github backend and would like to have the deploy preview links use the deployment from the Pull Request which we currently have it configured to automatically generate when the PR is updated.
